### PR TITLE
BUG: fix a crash when wrapping large integer Longitude

### LINKED ISSUE #16217: BUG: crash when instantiating a `Longitude` from large angle with `dtype=int`
### Description

We have at least one test showing that instantiating `Longitude` with an integer value is explicitly supported,
however it doesn't exercise wrapping, which is actually broken.
This came up while digging into #13479

### Expected behavior

No crash.
Alternatively, we could drop support for integer-valued angles altogether, but this seems uncalled for at this point.

### How to Reproduce

```python
import numpy as np
import astropy.units as u
from astropy.coordinates import Longitude

# this is explicitly tested in astropy/coordinates/tests/test_angles::test_longitude
assert Longitude(0, u.deg, dtype=int).dtype == np.dtype("int")

# but this crashes
Longitude(1234, u.deg, dtype=int).dtype
```

```python-traceback
Traceback (most recent call last):
  File "/Users/clm/dev/astropy-project/coordinated/astropy/untitled.py", line 9, in <module>
    Longitude(1234, u.deg, dtype=int).dtype
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/clm/dev/astropy-project/coordinated/astropy/astropy/coordinates/angles/core.py", line 716, in __new__
    self.wrap_angle = wrap_angle  # angle-like b/c property setter
    ^^^^^^^^^^^^^^^
  File "/Users/clm/dev/astropy-project/coordinated/astropy/astropy/coordinates/angles/core.py", line 733, in wrap_angle
    self._wrap_at(self.wrap_angle)
  File "/Users/clm/dev/astropy-project/coordinated/astropy/astropy/coordinates/angles/core.py", line 422, in _wrap_at
    self_angle -= wraps * a360
numpy._core._exceptions._UFuncOutputCastingError: Cannot cast ufunc 'subtract' output from dtype('float64') to dtype('int64') with casting rule 'same_kind'
```

### Versions

```python
import platform; print(platform.platform())
import sys; print("Python", sys.version)
import astropy; print("astropy", astropy.__version__)
import numpy; print("Numpy", numpy.__version__)
import erfa; print("pyerfa", erfa.__version__)
try:
    import scipy
    print("Scipy", scipy.__version__)
except ImportError:
    print("Scipy not installed")
try:
    import matplotlib
    print("Matplotlib", matplotlib.__version__)
except ImportError:
    print("Matplotlib not installed")
```
```
astropy 6.1.dev627+gce7673a9db.d20240319
Numpy 2.1.0.dev0+git20240318.6924f8b
pyerfa 2.0.1.2.dev19+g4a637c2
Scipy not installed
Matplotlib not installed
```

### PR DESCRIPTION
**Pull request by neutrinoceros**
### Description
Fixes #16217

Since this is a performance-critical function, I strongly recommend running benchmarks against this PR.

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

File: astropy/coordinates/angles/core.py [hunk 1]
```diff
@@ -404,7 +404,7 @@ def _wrap_at(self, wrap_angle):
         # than Quantity objects for speed.
         a360 = u.degree.to(self.unit, 360.0)
         wrap_angle = wrap_angle.to_value(self.unit)
-        self_angle = self.view(np.ndarray)
+        self_angle = self.view(np.ndarray).astype(f"f{self.dtype.itemsize}", copy=False)
```
Comment (neutrinoceros) [inline astropy/coordinates/angles/core.py:407]: originally I just cast to `float64` but then got failures in tests exercising `float32`, so this more complex casting ensures that preserve floating point precision, while still enabling int->float conversion

Comment (mhvk) [reply astropy/coordinates/angles/core.py:407]: This doesn't work - it has to be a view in order to do the in-place wrapping. I think we'll need to forbid integer.

Comment (neutrinoceros) [reply astropy/coordinates/angles/core.py:407]: good point. Let me wrap up an alternative PR then and see if it breaks more than one test

Comment (neutrinoceros) [reply astropy/coordinates/angles/core.py:407]: see #16220

**Comment by neutrinoceros** [comment]
As @mhvk, this doesn't work. It's also probably niche enough that documenting as a "known limitation" would be enough of a fix.

**Pull request closed without merging**